src/analyzer.ts

```diff
@@ -116,18 +123,66 @@ export class Analyzer {
     // re-analyze a particular file. Editors can use a UrlLoader that reads from
     // it's internal buffers.
     if (contents != null) {
+      this._scannedDocumentPromises.delete(resolvedUrl);
       this._scannedDocuments.delete(resolvedUrl);
-      this._parsedDocuments.delete(resolvedUrl);
+      this._parsedDocumentPromises.delete(resolvedUrl);
+      this._analyzedDocuments.delete(resolvedUrl);
+      this._analyzedDocumentPromises.delete(resolvedUrl);
+    }
+
+    const cachedResult = this._analyzedDocumentPromises.get(resolvedUrl);
+    if (cachedResult) {
+      return cachedResult;
     }
+    // TODO(justinfagnani): also check _analyzedDocuments?
+
+    const promise = (async() => {
+      // Make sure we wait and return a Promise before doing any work, so that
+      // the Promise is cached before anything else happens.
+      await Promise.resolve();
+      const scannedDocument = await this._scan(resolvedUrl, contents);
+      return this._makeDocument(scannedDocument);
+    })();
+    this._analyzedDocumentPromises.set(resolvedUrl, promise);
+    return promise;
+  }
 
-    const scannedDocument = await this._scan(resolvedUrl, contents);
+  /**
+   * Constructs a new analyzed Document and adds it to the analyzed Document
+   * cache.
+   */
+  private _makeDocument(scannedDocument: ScannedDocument): Document {
     const doneTiming =
-        this._telemetryTracker.start('Document.makeRootDocument', url);
-    const document = Document.makeRootDocument(scannedDocument);
+        this._telemetryTracker.start('_makeDocument', scannedDocument.url);
+    if (this._analyzedDocuments.has(scannedDocument.url)) {
+      throw new Error(
+          `Internal error: document ${scannedDocument.url} already exists`);
+    }
+    const document = new Document(scannedDocument, this);
+    document.resolve();
+    // TODO(justinfagnani): should this add to _analyzedDocumentPromises?
+    this._analyzedDocuments.set(scannedDocument.url, document);
     doneTiming();
     return document;
   }
 
+  /**
+   * Gets an analyzed Document from the document cache.
+   *
+   * If a document has been analyzed, it returns the analyzed Document. If not
+   * the scanned document cache is used and a new analyzed Document is returned.
+   * If a file is in neither cache, it returns `null`.
+   */
+  getDocument(url: string): Document {
```

Comment:
Can we make it available to Import.resolve somehow without exposing it to general users? I dislike having two APIs for the same thing, one that always works and the other than has caveats.

At the least I think we should add docs to this method that say "You almost certainly don't want to use this and should use `resolve` instead."